function-controller/pkg/controller/autoscaler/autoscaler.go

```diff
@@ -0,0 +1,325 @@
+			queueLen, err := a.transportInspector.QueueLength(topic, funcId.Function)
+			if err != nil {
+				log.Printf("Failed to obtain queue length (and will assume it is positive): %v", err)
+				return false, -1
+			}
+			if queueLen > 0 {
+				return false, queueLen
+			}
+		}
+	}
+	return true, 0
+}
+
+func (a *autoScaler) StartMonitoring(topic string, fn FunctionId) error {
+	a.mutex.Lock()
+	defer a.mutex.Unlock()
+	funcTotals, ok := a.totals[topic]
+	if !ok {
+		funcTotals = make(map[FunctionId]*metricsTotals)
+		a.totals[topic] = funcTotals
+	}
+
+	_, ok = funcTotals[fn]
+	if ok {
+		return fmt.Errorf("Already monitoring topic %s and function %s", topic, fn)
+	}
+
+	funcTotals[fn] = &metricsTotals{}
+
+	a.scalers[fn] = decorate(a.metricsScaler(fn), a.limitScalingUp(fn), a.limitScalingDown(fn), a.delay(fn))
+
+	return nil
+}
+
+func (a *autoScaler) delay(fn FunctionId) adjuster {
+	p := NewDelayer(func() time.Duration {
+		return a.delayScaleDown(fn);
+	})
+
+	return func(proposal int) int {
+		return p.Delay(proposal).Get()
+	}
+}
+
+func (a *autoScaler) StopMonitoring(topic string, function FunctionId) error {
+	a.mutex.Lock()
+	defer a.mutex.Unlock()
+	funcTotals, ok := a.totals[topic]
+	if !ok {
+		return fmt.Errorf("Not monitoring topic %s and function %s", topic, function)
+	}
+
+	_, ok = funcTotals[function]
+	if !ok {
+		return fmt.Errorf("Not monitoring topic %s and function %s", topic, function)
+	}
+
+	delete(funcTotals, function)
+
+	// Avoid leaking memory.
+	if len(funcTotals) == 0 {
+		delete(a.totals, topic)
+	}
+	delete(a.scalers, function)
+
+	return nil
+}
+
+func (a *autoScaler) receiveLoop() {
+	producerMetrics := a.metricsReceiver.ProducerMetrics()
+	consumerMetrics := a.metricsReceiver.ConsumerMetrics()
+	for {
+		select {
+		case pm, ok := <-producerMetrics:
+			if ok {
+				a.receiveProducerMetric(pm)
+			}
+
+		case cm, ok := <-consumerMetrics:
+			if ok {
+				a.receiveConsumerMetric(cm)
+			}
+
+		case <-a.stop:
+			close(a.accumulatingStopped)
+			return
+		}
+	}
+}
+
+func (a *autoScaler) receiveConsumerMetric(cm metrics.ConsumerAggregateMetric) {
+	a.mutex.Lock()
+	defer a.mutex.Unlock()
+
+	funcTotals, ok := a.totals[cm.Topic]
+	if ok {
+		mt, ok := funcTotals[FunctionId{cm.ConsumerGroup}]
+		if ok {
+			mt.receiveCount += cm.Count
+		}
+	}
+}
+
+func (a *autoScaler) receiveProducerMetric(pm metrics.ProducerAggregateMetric) {
+	a.mutex.Lock()
+	defer a.mutex.Unlock()
+
+	funcTotals, ok := a.totals[pm.Topic]
+	if ok {
+		for _, mt := range funcTotals {
+			mt.transmitCount += pm.Count
+		}
+	}
+}
+
+type scaler func(*metricsTotals) int
+type adjuster func(int) int
+
+func decorate(s scaler, adjusters ...adjuster) scaler {
+	for _, a := range adjusters {
+		s = compose(a, s)
+	}
+
+	return s
+}
+
+func compose(a adjuster, s scaler) scaler {
+	return func(mt *metricsTotals) int {
+		return a(s(mt))
+	}
+}
+
+func (a *autoScaler) metricsScaler(fn FunctionId) func(mt *metricsTotals) int {
```

Comment:
return type can be changed to `scaler`

Reply:
Done. Also changes relevant function return types to `adjuster` similarly.